lib/ansible/modules/network/cloudengine/ce_ntp.py

```diff
@@ -0,0 +1,621 @@
+            self.address = self.peer
+        else:
+            self.peer_type = None
+            self.address = None
+
+        self.check_params()
+
+        # state
+        self.changed = False
+        self.updates_cmd = list()
+        self.results = dict()
+        self.proposed = dict()
+        self.existing = list()
+        self.end_state = list()
+
+        self.init_data()
+
+    def init_data(self):
+        """Init data"""
+
+        if self.interface is not None:
+            self.interface = self.interface.lower()
+
+        if not self.key_id:
+            self.key_id = ""
+
+        if not self.is_preferred:
+            self.is_preferred = 'disable'
+
+    def init_module(self):
+        """Init module"""
+
+        self.module = AnsibleModule(argument_spec=self.spec, supports_check_mode=True,
+                                    mutually_exclusive=self.mutually_exclusive)
+
+    def check_ipaddr_validate(self):
+        """Check ipaddress validate"""
+
+        rule1 = r'(25[0-5]|2[0-4][0-9]|1[0-9][0-9]|[1-9]?[0-9])\.'
+        rule2 = r'(25[0-5]|2[0-4][0-9]|1[0-9][0-9]|[1-9]?[0-9])'
+        ipv4_regex = '%s%s%s%s%s%s' % ('^', rule1, rule1, rule1, rule2, '$')
+        ipv6_regex = '^(?:[a-fA-F0-9]{1,4}:){7}[a-fA-F0-9]{1,4}$'
+
+        flag = False
+        if bool(re.match(ipv4_regex, self.address)):
+            flag = True
+            self.ip_ver = "IPv4"
+            if not self.ntp_ucast_ipv4_validate():
+                flag = False
+        elif bool(re.match(ipv6_regex, self.address)):
+            flag = True
+            self.ip_ver = "IPv6"
+        else:
+            flag = True
+            self.ip_ver = "IPv6"
+
+        if not flag:
+            if self.peer_type == "Server":
+                self.module.fail_json(msg='Error: Illegal server ip-address.')
+            else:
+                self.module.fail_json(msg='Error: Illegal peer ip-address.')
+
+    def ntp_ucast_ipv4_validate(self):
+        """Check ntp ucast ipv4 address"""
+
+        addr_list = re.findall(r'(.*)\.(.*)\.(.*)\.(.*)', self.address)
+        if not addr_list:
+            self.module.fail_json(msg='Error: Match ip-address fail.')
+
+        value = ((long(addr_list[0][0])) * 0x1000000) + (long(addr_list[0][1])\
+                * 0x10000) + (long(addr_list[0][2]) * 0x100) + (long(addr_list[0][3]))
+        if (value & (0xff000000) == 0x7f000000) or (value & (0xF0000000) == 0xF0000000) \
+                or (value & (0xF0000000) == 0xE0000000) or (value == 0):
+            return False
+        return True
+
+    def check_params(self):
+        """Check all input params"""
+
+        # check interface type
+        if self.interface:
+            intf_type = get_interface_type(self.interface)
+            if not intf_type:
+                self.module.fail_json(
+                    msg='Error: Interface name of %s '
+                        'is error.' % self.interface)
+
+        if not self.server and not self.peer:
```

Comment:
There are some options to `AnsibleModule` which you may be able to use to validate options passed in.
Look at existing modules for examples:
```
required_one_of
```